source/simulator/solver_schemes.cc

```diff
@@ -0,0 +1,744 @@
+                                                         double *initial_residual)
+  {
+    assemble_advection_system (AdvectionField::temperature());
+
+    if (compute_initial_residual)
+      {
+        Assert(initial_residual != NULL, ExcNotImplemented());
+        *initial_residual = system_rhs.block(introspection.block_indices.temperature).l2_norm();
+      }
+
+    const double current_residual = solve_advection(AdvectionField::temperature());
+    current_linearization_point.block(introspection.block_indices.temperature)
+      = solution.block(introspection.block_indices.temperature);
+
+    if ((initial_residual != NULL) && (*initial_residual > 0))
+      return current_residual / *initial_residual;
+
+    return 0.0;
+  }
+
+  template <int dim>
+  std::vector<double> Simulator<dim>::assemble_and_solve_composition (const bool compute_initial_residual,
+                                                                      std::vector<double> *initial_residual)
+  {
+    std::vector<double> current_residual(introspection.n_compositional_fields,0.0);
+
+    if (compute_initial_residual)
+      Assert(initial_residual->size() == introspection.n_compositional_fields, ExcNotImplemented());
+
+    for (unsigned int c=0; c < introspection.n_compositional_fields; ++c)
+      {
+        const AdvectionField adv_field (AdvectionField::composition(c));
+        const typename Parameters<dim>::AdvectionFieldMethod::Kind method = adv_field.advection_method(introspection);
+        switch (method)
+          {
+            case Parameters<dim>::AdvectionFieldMethod::fem_field:
+              assemble_advection_system (adv_field);
+
+              if (compute_initial_residual)
+                (*initial_residual)[c] = system_rhs.block(introspection.block_indices.compositional_fields[c]).l2_norm();
+
+              current_residual[c] = solve_advection(adv_field);
+              break;
+
+            case Parameters<dim>::AdvectionFieldMethod::particles:
+              interpolate_particle_properties(adv_field);
+              break;
+
+            default:
+              AssertThrow(false,ExcNotImplemented());
+          }
+      }
+
+    // for consistency we update the current linearization point only after we have solved
+    // all fields, so that we use the same point in time for every field when solving
+    for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+      {
+        current_linearization_point.block(introspection.block_indices.compositional_fields[c])
+          = solution.block(introspection.block_indices.compositional_fields[c]);
+
+        if ((initial_residual != NULL) && (*initial_residual)[c] > 0)
+          current_residual[c] /= (*initial_residual)[c];
+        else
+          current_residual[c] = 0.0;
+      }
+
+    return current_residual;
+  }
+
+
+
+  template <int dim>
+  double Simulator<dim>::assemble_and_solve_stokes (const bool compute_initial_residual,
```

Comment:
I think you use uppercase `Stokes` for some of these functions, but this one is lowercase `stokes`. Maybe make it uniform?

Reply:
I think we tended to use lower-case `stokes` everywhere. Maybe that suggests also using `newton_stokes`?